M2/Macaulay2/packages/SimpleDoc.m2

```diff
@@ -1,332 +1,291 @@
-singleString = (key, textlines, keylinenum) -> (
-     if #textlines === 0 then
-       error("line ",toString keylinenum," of string: expected single indented line after "|toString key)
-     else if #textlines > 1 then 
-       error("line ",toString getLinenum textlines#1," of string: expected single indented line after "|toString key);
-     key => getText textlines#0)
-
-multiString = (key, textlines, keylinenum) -> (				    -- written by Andrew Hoefel originally
-     if #textlines === 0 then
-       error("line ",toString keylinenum," of string: expected at least one indented line after "|toString key);
-     key => concatenate between(newline, getText \ textlines)
-     )
-
-listOfStrings = (key, textlines, keylinenum) -> (
-     key => getText \ textlines
-     )
-
-reassemble = (indent,textlines) -> concatenate between(newline,
-     for x in textlines list ( if getIndent x =!= infinity then getIndent x - indent : " ", getText x )
-     )
-
-safevalue = t -> (
-     try value t else (
-	  stderr << "in the evaluation of: " << stack lines t << endl;
-	  value t
-	  )
-     )
-
-markup2 = (textlines, keylinenum) -> (
-     if #textlines === 0 then return "";
-     s := concatenate between(" ",getText \ textlines);
-     sp := separateRegexp(///(^|[^\\])(@)///, 2, s);
-     sp = apply(sp, s -> replace(///\\@///,"@",s));
-     if not odd(#sp) then error "unmatched @";
-     t := for i from 0 to #sp-1 list if even i then if sp#i != "" then TEX sp#i else "" else safevalue concatenate("(",sp#i,")");
-     t = select(t, x -> x =!= "");
-     if instance(t,List) and #t === 1 then t = first t;
-     t
-     )
+    "SimpleDoc",
+    Version => "1.2",
+    Date => "May 18, 2020",
+    Headline => "a simple documentation function",
+    Authors => {
+	{ Name => "Dan Grayson", Email => "dan@math.uiuc.edu", HomePage => "https://faculty.math.illinois.edu/~dan/" },
+	{ Name => "Mike Stillman", Email => "mike@math.cornell.edu", HomePage => "https://pi.math.cornell.edu/~mike/" },
+	{ Name => "Mahrud Sayrafi", Email => "mahrud@umn.edu", HomePage => "https://math.umn.edu/~mahrud/" }
+	},
+    PackageImports => { "Text" },
+    DebuggingMode => false,
+    AuxiliaryFiles => true
+    )
+
+export {"doc", "multidoc", "packageTemplate", -- functions
+    "arXiv", "stacksProject", "wikipedia", -- helper functions
+    "docTemplate", "docExample", "testExample", "simpleDocFrob" -- templates and examples
+    }
+
+-- The class of processed documentation nodes
+Node = new IntermediateMarkUpType of Hypertext
+Node.synonym = "processed documentation node"
+
+-- Primary functions
+doc = method()
+doc String := str -> (
+    docstring := if fileExists str then get str else str;
+    parsed := toDoc(NodeFunctions, docstring);
+    document \ (
+	if all(parsed, elt -> instance(elt, Node)) then apply(parsed, node -> toList node)
+	else if not any(parsed, elt -> instance(elt, Node)) then {parsed}
+	else error("expected either a single or a list of documentation nodes")))
```

Comment:
This function now returns a list of nulls, but it shouldn't.  (Use `scan` instead of `apply`.)
```m2
Macaulay2, version 1.16.0.2
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               MinimalPrimes, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : doc ///
     Key
      a
     ///

o1 = {}

o1 : List

i2 : peek oo

o2 = {null}
```

Reply:
Ah, this is intentional, but it'll make more sense in the next PR, where `document` will actually return the nodes so you can inspect them directly with `show` and other commands. I'll change it for now.